overlord/configstate/configcore/netplan.go

```diff
@@ -92,18 +118,147 @@ func getNetplanFromSystem(key string) (result interface{}, err error) {
 	netplan := conn.Object("io.netplan.Netplan", "/io/netplan/Netplan")
 
 	if err := netplan.Call("io.netplan.Netplan.Config", 0).Store(&netplanConfigSnapshotBusAddr); err != nil {
-		if derr, ok := err.(dbus.Error); ok {
-			// Having no netplan config is *not* an error, we just
-			// do not support netplan config.
-			if isNoServiceOrMethodErr(derr) {
-				return nil, nil
-			}
+		return nil, err
+	}
+	logger.Debugf("using netplan config snapshot %v", netplanConfigSnapshotBusAddr)
+
+	netplanCfgSnapshot := conn.Object("io.netplan.Netplan", dbus.ObjectPath(netplanConfigSnapshotBusAddr))
+	return netplanCfgSnapshot, nil
+}
+
+func validateNetplanSettings(tr config.Conf) error {
+	// XXX: validate somehow once we support writing?
+	return nil
+}
+
+func isNetplanChange(chg string) bool {
+	return chg == "core.system.network.netplan" || strings.HasPrefix(chg, "core.system.network.netplan.")
+}
+
+func hasNetplanChanges(tr config.Conf) bool {
+	for _, chg := range tr.Changes() {
+		if isNetplanChange(chg) {
+			return true
+		}
+	}
+	return false
+}
+
+var storeReachableRetryWait = 1 * time.Second
+
+func handleNetplanConfiguration(tr config.Conf, opts *fsOnlyContext) error {
+	if !hasNetplanChanges(tr) {
+		return nil
+	}
+
+	var cfg map[string]interface{}
+	if err := tr.Get("core", "system.network.netplan", &cfg); err != nil && !config.IsNoOption(err) {
+		return err
+	}
+
+	netplanCfgSnapshot, err := getNetplanCfgSnapshot()
+	// Having no netplan config is *not* an error, we just
+	// do not support netplan config.
+	if isNoServiceOrMethodErr(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// Use the default config that most Ubuntu Core have
+	originHint := "90-snapd-config"
+
+	// Always starts with a clean config to avoid merging of keys
+	// that got unset.
+	configs := []string{"network=null"}
+	// and then pass the full new config in
+	for key := range cfg {
+		// We pass the new config back to netplan as json, the reason
+		// is that the dbus api accepts only a single line string, see
+		// see https://github.com/canonical/netplan/pull/210
+		jsonNetplanConfigRaw, err := json.Marshal(cfg[key])
+		if err != nil {
+			return fmt.Errorf("cannot netplan config: %v", err)
+		}
+		configs = append(configs, fmt.Sprintf("%s=%s", key, string(jsonNetplanConfigRaw)))
+	}
+
+	// now apply
+	for _, jsonNetplanConfig := range configs {
+		logger.Debugf("calling netplan.Set: %v", jsonNetplanConfig)
```

Comment:
calling netplan.Set with: ?

Reply:
Thanks, I decided to remove this debug line, it is useful but the risk is that it logs wifi (or vpn) credentials so it feels safer to not have this level of detailed logging.